src/uri.rs

```diff
@@ -0,0 +1,261 @@
+    // item stats supporst 3 end points/
+    // base /itemstats
+    // id   /itemstats/id
+    // multiple(?) /itemstats?ids=23
+    item_stats_all(Option<ItemStatId>),
+    recipes(Option<RecipeId>),
+    build,
+}
+
+impl EndPoint {
+    pub fn requires_auth(self) -> bool {
+        match self {
+            // do require auth.
+            EndPoint::account => true,
+            EndPoint::account_achievements => true,
+            EndPoint::account_bank => true,
+            EndPoint::account_materials => true,
+            EndPoint::account_dailycrafting => true,
+            EndPoint::account_dungeons => true,
+            EndPoint::account_dyes => true,
+
+            // don't require auth
+            EndPoint::achievements(_) => false,
+            EndPoint::achievements_daily => false,
+            EndPoint::achievements_daily_tomorrow => false,
+            EndPoint::items(_) => false,
+            EndPoint::item_stats(_) => false,
+            EndPoint::item_stats_all(_) => false,
+            EndPoint::recipes(_) => false,
+            EndPoint::build => false,
+        }
+    }
+
+    pub fn uri(&self) -> String {
+        match self {
+            EndPoint::account => format!("account"),
+            EndPoint::account_achievements => format!("account/achievements"),
+            EndPoint::account_dailycrafting => format!("account/dailycrafting"),
+            EndPoint::account_dungeons => format!("account/dungeons"),
+            EndPoint::account_dyes => format!("account/dyes"),
+            EndPoint::achievements(op_id) => match op_id {
+                Some(id) => format!("achievements/{}", id.0.to_string()),
+                None => format!("achievements"),
+            },
+            EndPoint::achievements_daily => "achievements/daily".to_string(),
+            EndPoint::achievements_daily_tomorrow => "achievements/daily/tomorrow".to_string(),
+            EndPoint::account_materials => "account/materials".to_string(),
+            EndPoint::account_bank => "account/bank".to_string(),
+            EndPoint::items(id) => format!("items/{}", id.0.to_string()),
+            EndPoint::item_stats(op_stats_id) => match op_stats_id {
+                Some(stats_id) => format!("itemstats/{}", stats_id.0.to_string()),
+                None => "itemstats".to_string(),
+            },
+            EndPoint::item_stats_all(op_item_stat_id) => match op_item_stat_id {
+                Some(item_stat_id) => format!("itemstats/{}", item_stat_id.0.to_string()),
+                None => "itemstats".to_string(),
+            },
+            EndPoint::recipes(op_recipe_id) => match op_recipe_id {
+                Some(recipe_id) => format!("recipes/{}", recipe_id.0.to_string()),
+                None => "recipes".to_string(),
+            },
+            EndPoint::build => "build".to_string(),
+        }
+    }
+}
+
+pub struct Requester {
+    version: ApiVersion,
+    api_key: Option<ApiKey>,
```

Comment:
This attribute is currently unused.

Also I have some use cases where I will be requesting data from many different users.  Making a Requester per api key may be kinda odd.  Really depends on the goal of this type.

If a publicly facing (Rust) API requires an API key I'd prefer that it explicitly takes it as a parameter (and also documents the permission requirements).  An optional API key in a requester seems like we will be making requesters per request (which eliminates the potential value add of injecting the base_uri) or we will be sneaking a Option\<ApiKey\> through all of our code, potentially doing the wrong thing.

Reply:
Could you expand on how have a Reqeuster per api key seems odd? That seem normal to me, as the binding of a Requester to the api key allows for building uris for that specific key. 

I picture a sort of API builder that constructs apis based on conveniently provided input from the user, having to have the user specify their api key on each request sounds cumbersome to me, hence the key can be stored within the Requester type. 

Its ment as optional due to not all the api end points require authentication and thus this is to allow for a user to request api end points that do not require authentication.

Sounds like you are suggesting to go towards a compile time check of api key and its permissions per end point? That sounds good me. Not certain what you mean by potentially doing the wrong thing though.